addons/stock/models/stock_location.py

```diff
@@ -176,23 +176,36 @@ def _get_putaway_strategy(self, product, quantity=0, package=None):
             putaway_rules |= self.putaway_rule_ids.filtered(lambda x: not x.product_id and (package_type in x.package_type_ids or package_type == x.package_type_ids))
 
         # get current product qty (qty in current quants and future qty on assigned ml) of all child locations
-        qty_by_location = {}
+        data_by_location = defaultdict(lambda: {'qty': 0, 'weight': 0})
         locations = self.env['stock.location'].search([('id', 'child_of', self.id), ('usage', '=', 'internal')])
         if locations.storage_category_id:
             move_line_data = self.env['stock.move.line'].read_group([
-                ('product_id', '=', product.id),
                 ('location_dest_id', 'in', locations.ids),
                 ('state', 'not in', ['draft', 'done', 'cancel'])
-            ], ['location_dest_id', 'product_qty:sum'], ['location_dest_id'])
+            ], ['location_dest_id', 'product_id', 'product_qty:sum'], ['location_dest_id', 'product_id'], lazy=False)
             quant_data = self.env['stock.quant'].read_group([
-                ('product_id', '=', product.id),
                 ('location_id', 'in', locations.ids),
-            ], ['location_id', 'quantity:sum'], ['location_id'])
+            ], ['location_id', 'product_id', 'quantity:sum'], ['location_id', 'product_id'], lazy=False)
+
+            # get weight of all related products
+            weight_by_product = defaultdict(lambda: 0)
+            product_ids = set([product.id])
+            for values in move_line_data + quant_data:
+                product_ids.add(values['product_id'][0])
+                products_weight = product.search_read(domain=[('id', 'in', list(product_ids))], fields=['id', 'weight'])
```

Comment:
I think you could move the search_read outside the loop to do only one search on all the products

Reply:
oh, huge mistake